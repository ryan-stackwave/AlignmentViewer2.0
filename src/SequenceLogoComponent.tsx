/**
 *
 * Sequence logo component.
 * Inspired / derived from https://github.com/weng-lab/logojs-package
 *  (but much simpler)
 */

import React from "react";
import Alignment from "./Alignment";
import { GlyphFactory } from "./SequenceLogoGlyphs";
import { acePrefix } from "./MolecularStyles";

export enum LOGO_TYPES {
  LETTERS = "letter stack",
  BARS = "bar plot"
}

interface IGlyphFrequency {
  frequency: number;
  letter: {
    letter: string;
    classNames: string;
  };
}
interface IGlyphStack extends Array<IGlyphFrequency> {}

export interface ISequenceLogoComponentProps {
  alignment: Alignment;
  glyphWidth: number;
  logoType: LOGO_TYPES;

  logoLoaded(node: HTMLDivElement): void;

  readonly id: string;
}

export class SequenceLogoComponent extends React.Component<
  ISequenceLogoComponentProps
> {
  private scrollerDiv?: HTMLDivElement | null;
  private loaded: boolean = false;

  divLoaded(e: HTMLDivElement | null) {
    if (e && !this.scrollerDiv) {
      //only run if first time element available
      this.scrollerDiv = e;
      this.props.logoLoaded(this.scrollerDiv);
    }
  }

  shouldComponentUpdate(nextProps: ISequenceLogoComponentProps) {
    if (
      this.loaded === false ||
      this.props.logoType !== nextProps.logoType ||
      this.props.glyphWidth !== nextProps.glyphWidth
    ) {
      return true;
    }
    return false;
  }

  render() {
    if (!this.props.alignment || !this.props.glyphWidth) {
      return null;
    }

    const positionalLetterCounts = this.props.alignment.getPositionalLetterCounts();
    const lettersSorted = this.props.alignment.getSortedAlphaLetters();

    //calculate the frequencies of all letters in each column
    const frequencies: number[][] = [];
    positionalLetterCounts.forEach((letterCounts, position) => {
      const totalCountAtPosition = Object.keys(letterCounts).reduce(
        (total, letter) => {
          total += letterCounts[letter];
          return total;
        },
        0
      );

      const frequenciesAtPosition: number[] = lettersSorted.map(letter => {
        let freq = 0;
        if (letter in letterCounts) {
          freq = letterCounts[letter] / totalCountAtPosition;
        }
        return freq;
      });

      frequencies.push(frequenciesAtPosition);
    });

    //load alphabet information
    const sortedAlphabet = lettersSorted.map(letter => {
      return {
        letter: letter,
        classNames: acePrefix + letter
      };
    });

    //munge in alphabet and sort by frequency
    const logoData = frequencies.map((frequenciesAtPosition, positionIdx) => {
      return frequenciesAtPosition
        .map((frequency, letterIdx) => {
          return {
            frequency: frequency,
            letter: sortedAlphabet[letterIdx]
          };
        })
        .filter(freq => {
          return freq.frequency !== 0;
        })
        .sort((a, b) => (a.frequency > b.frequency ? 1 : -1));
    });

    const totalWidth = frequencies.length * this.props.glyphWidth;
    const totalHeight = 100;
    this.loaded = true;

    /**
     * Generate the svg elements for a single position, i.e., column
     *
     */
    const generateStack = (positionalFrequencies: IGlyphStack) => {
      let dy = 100;

      const xscale = 2 / logoData.length; // not exact, but works.
      return positionalFrequencies.map((freq, idx) => {
        dy = dy - freq.frequency * 100;

        if (this.props.logoType === LOGO_TYPES.BARS) {
          return (
            <rect
              width="100"
              height="100"
              transform={`translate(0, ${dy}) scale(${xscale},${freq.frequency})`}
              className={freq.letter.classNames}
              key={`idxrect_${idx}`}
            ></rect>
          );
        }

        const selectedGlyph = GlyphFactory.glyphFromChar(freq.letter.letter)({
          className: freq.letter.classNames,
          transform: `translate(0, ${dy}) scale(${xscale},${freq.frequency})`,
          key: `idxglyph_${idx}`
        });
        return selectedGlyph;
      });
    };

    return (
      <div id={this.props.id} ref={e => this.divLoaded(e)}>
        <svg
          preserveAspectRatio="none"
          viewBox={`0 0 ${logoData.length} 100`}
          style={{ width: totalWidth, height: totalHeight }}
          xmlns="http://www.w3.org/2000/svg"
        >
          {logoData.map((singlePositionData, positionIdx) => {
            return (
              <g
                transform={`translate(${positionIdx},0)`}
                key={"p_" + positionIdx}
              >
                {generateStack(singlePositionData)}
              </g>
            );
          })}
        </svg>
      </div>
    );
  }
}
