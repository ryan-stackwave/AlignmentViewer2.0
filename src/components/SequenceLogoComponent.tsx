/**
 * Sequence logo component.
 * Inspired / derived from https://github.com/weng-lab/logojs-package
 *  (but simpler)
 */

import React from "react";
import "./SequenceLogo.scss";
import { Alignment } from "../common/Alignment";
import { GlyphFactory } from "../common/SequenceLogoGlyphs";
import {
  aceResidueParentClass,
  getLetterClassNames,
  AlignmentTypes,
} from "../common/MolecularStyles";
import ReactTooltip from "react-tooltip";
import { AminoAcid, Nucleotide } from "../common/Residues";

export enum LOGO_TYPES {
  LETTERS = "Letter Stack",
  BARS = "Bar Plot",
}

interface ILetterWithClasses {
  letter: string;
  classNames: string;
}
interface IGlyphFrequency {
  count: number;
  frequency: number;
  letter: ILetterWithClasses;
  residue: AminoAcid | Nucleotide;
}
interface IGlyphStackData extends Array<IGlyphFrequency> {}

export interface ISequenceLogoProps {
  alignment: Alignment;
  glyphWidth: number;
  height: number;
  logoType: LOGO_TYPES;
  alignmentType?: AlignmentTypes;
}

interface ISequenceLogoState {
  logoData?: IGlyphStackData;
}

export class SequenceLogoComponent extends React.Component<
  ISequenceLogoProps,
  ISequenceLogoState
> {
  private logoData?: IGlyphStackData[];

  static defaultProps = {
    logoType: LOGO_TYPES.LETTERS,
    height: 100,
  };

  /*
   *
   *
   * PRIVATE METHODS
   *
   *
   */

  private renderTooltip() {
    return (
      <ReactTooltip
        id="getLogoTooltip"
        effect="solid"
        type="light"
        border={true}
        getContent={(idx: string) => {
          if (!idx || !this.logoData) {
            return;
          }

          const position = parseInt(idx) + 1;
          const glyphData = this.logoData![parseInt(idx)];
          return glyphData.length === 0 ? null : (
            <div className="logo-tooltip">
              <h1>Position: {position}</h1>
              <div className="row">
                <div className="col legend-square"></div>
                <div className="col header name">Residue</div>
                <div className="col header frequency">Frequency</div>
                <div className="col header count">Count</div>
              </div>
              {glyphData
                .slice()
                .reverse()
                .map((letterFreq, myidx) => {
                  return (
                    <div key={myidx} className="row">
                      <div
                        className={`col legend-square ${letterFreq.letter.classNames}`}
                      ></div>
                      <div className="col name">
                        {`${letterFreq.residue.fullName} [${letterFreq.letter.letter}]`}
                      </div>
                      <div className="col frequency">
                        {+(letterFreq.frequency * 100).toFixed(2)}%
                      </div>
                      <div className="col count">{letterFreq.count}</div>
                    </div>
                  );
                })}
            </div>
          );
        }}
      />
    );
  }

  /**
   * Generate the svg elements for a single position, i.e., column
   */
  private renderSinglePositionStack(
    positionalFrequencies: IGlyphStackData,
    numGlyphStacks: number
  ) {
    const { logoType, height } = this.props;
    let dy = height;

    return positionalFrequencies.map((freq, idx) => {
      dy = dy - freq.frequency * 100;

      if (logoType === LOGO_TYPES.BARS) {
        const xscale = 2 / numGlyphStacks; // not exact, but works.
        return (
          <rect
            width="100"
            height="100"
            transform={`translate(0, ${dy}) scale(${xscale},${freq.frequency})`}
            className={freq.letter.classNames}
            key={`idxrect_${idx}`}
          ></rect>
        );
      }

      const xscale = 0.009; // invariant scaling
      const selectedGlyph = GlyphFactory.glyphFromChar(freq.letter.letter)({
        className: freq.letter.classNames,
        transform: `translate(0, ${dy}) scale(${xscale},${freq.frequency})`,
        key: `idxglyph_${idx}`,
      });
      return selectedGlyph;
    });
  }

  /**
   * Munge letter count data that was calculated during alignment creation
   * into a form appropriate for the glyph generation
   */
  private mungeLogoData(): IGlyphStackData[] {
    const { alignment } = this.props;

    const numberSequences = alignment.getSequences().length;
    const lettersSorted = alignment.getAllUpperAlphaLettersInAlignmentSorted();

    const alignmentType = this.props.alignmentType
      ? this.props.alignmentType
      : this.props.alignment.getPredictedType();
    const moleculeClass =
      alignmentType === AlignmentTypes.AMINOACID ? AminoAcid : Nucleotide;

    //load class names for each letter
    const letterObjects = lettersSorted.reduce((arr, letter) => {
      arr[letter] = {
        letter: letter,
        classNames: getLetterClassNames(letter, false, false, false),
      };
      return arr;
    }, {} as { [letter: string]: ILetterWithClasses });

    //note: removes invalid letters, but letterCount (value) isn't sorted
    const plc = alignment.getPositionalLetterCounts(false, lettersSorted);
    return Array.from(plc).map(
      ([pos, letterCounts]): IGlyphStackData => {
        return Object.entries(letterCounts)
          .map(([letter, count]) => {
            return {
              count: count,
              frequency: count / numberSequences,
              letter: letterObjects[letter],
              residue: moleculeClass.fromSingleLetterCode(letter),
            };
          })
          .filter((freq) => {
            // remove empty letters
            return freq.frequency !== 0;
          })
          .sort((a, b) => (a.frequency > b.frequency ? 1 : -1));
      }
    );
  }

  /*
   *
   *
   * REACT METHODS
   *
   *
   */
  shouldComponentUpdate(nextProps: ISequenceLogoProps) {
    const { alignment, alignmentType, logoType, glyphWidth } = this.props;
    if (
      !this.logoData ||
      logoType !== nextProps.logoType ||
      glyphWidth !== nextProps.glyphWidth ||
      alignment !== nextProps.alignment ||
      alignmentType !== nextProps.alignmentType
    ) {
      return true;
    }
    return false;
  }

  render() {
    const { alignment, glyphWidth, height } = this.props;
    if (!alignment || !glyphWidth) {
      return null;
    }

    //perform a bunch of data munging
    this.logoData = this.mungeLogoData();
    const totalWidth = this.logoData.length * glyphWidth;
    const totalHeight = height;

    return (
      <div className="sequence-logo">
        <svg
          preserveAspectRatio="none"
          viewBox={`0 0 ${this.logoData.length} 100`}
          style={{ width: totalWidth, height: totalHeight }}
          xmlns="http://www.w3.org/2000/svg"
        >
          {this.logoData.map((singlePositionData, positionIdx) => {
            return (
              <g
                transform={`translate(${positionIdx},0)`}
                className={aceResidueParentClass} //required for default coloring
                key={"p_" + positionIdx}
              >
                {this.renderSinglePositionStack(
                  singlePositionData,
                  this.logoData!.length
                )}
                <rect
                  className="interaction-placeholder"
                  width="1"
                  height="100"
                  data-for="getLogoTooltip"
                  data-tip={positionIdx}
                  data-class={"sequence-logo-tooltip-container"}
                ></rect>
              </g>
            );
          })}
        </svg>
        {this.renderTooltip()}
      </div>
    );
  }
}
