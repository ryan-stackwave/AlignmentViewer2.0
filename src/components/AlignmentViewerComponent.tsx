import React from "react";
import "./AlignmentViewer.scss";
import { Ace } from "ace-builds";
import Alignment, { SequenceSortOptions } from "../common/Alignment";
import ScrollSync, { ScrollType } from "../common/ScrollSync";
import { SequenceLogoComponent, LOGO_TYPES } from "./SequenceLogoComponent";
import { SequenceConservationComponent } from "./SequenceConservationComponent";
import {
  AminoAcidAlignmentStyle,
  NucleotideAlignmentStyle,
  PositionsToStyle,
} from "../common/MolecularStyles";
import MiniMapComponent from "./MiniMapComponent";
import AceMultipleSequenceAlignmentComponent from "./AceMultipleSequenceAlignmentComponent";
import AceConsensusSequenceComponent from "./AceConsensusSequenceComponent";
import AceTargetSequenceComponent from "./AceTargetSequenceComponent";
import AceTextualRulerComponent from "./AceTextualRulerComponent";
import { AceEditorComponent } from "./AceEditorComponent";

export interface IAlignmentViewerProps {
  alignment: Alignment;
  style: AminoAcidAlignmentStyle | NucleotideAlignmentStyle;
  logoPlotStyle: LOGO_TYPES;
  zoomLevel: number;
  sortBy: SequenceSortOptions;
  showMiniMap?: boolean;
  showAnnotations?: boolean;
}
export interface IAlignmentViewerState {
  aceCharacterWidth: number;
  aceEditors: Ace.Editor[];
  alignmentEditorVisibleFirstRow?: number;
  alignmentEditorVisibleLastRow?: number;

  windowWidth: number;
  windowHeight: number;
}

export class AlignmentViewer extends React.Component<
  IAlignmentViewerProps,
  IAlignmentViewerState
> {
  constructor(props: IAlignmentViewerProps) {
    super(props);
    this.state = {
      aceEditors: [],
      aceCharacterWidth: 0,
      windowWidth: window.innerWidth,
      windowHeight: window.innerHeight,
    };

    //setup scroll groups
    ScrollSync.getInstance().setScrollerGroup(
      "horizontal",
      ScrollType.horizontal
    );
    ScrollSync.getInstance().setScrollerGroup("vertical", ScrollType.vertical);
  }

  private handleCharacterSizeChanged = (newCharSize: number) => {
    if (this.state.aceCharacterWidth !== newCharSize) {
      this.setState({
        aceCharacterWidth: newCharSize,
      });
    }
  };

  private aceEditorLoaded = (
    id: string,
    editor: Ace.Editor,
    scrollSyncDirection: ScrollType
  ) => {
    //console.log("_aceEditorLoaded id =" + id);

    let scrollSync = ScrollSync.getInstance();
    if (
      scrollSyncDirection === ScrollType.horizontal ||
      scrollSyncDirection === ScrollType.both
    ) {
      scrollSync.registerScroller(editor, "horizontal");
    }
    if (
      scrollSyncDirection === ScrollType.vertical ||
      scrollSyncDirection === ScrollType.both
    ) {
      scrollSync.registerScroller(editor, "vertical");
    }
    this.setState({
      aceCharacterWidth: editor.renderer.characterWidth, //todo: check if the same always.
      aceEditors: [editor].concat(this.state.aceEditors),
    });

    //track visible rows to show in canvas MSA
    if (id === "ace-alignment") {
      editor.renderer.on("afterRender", () => {
        // BREAKS LOGO, rerenders everything below and kills performance. React lifecycle stuff.
        setTimeout(() => {
          this.setState({
            alignmentEditorVisibleFirstRow: editor.renderer.getFirstFullyVisibleRow(),
            alignmentEditorVisibleLastRow: editor.renderer.getLastFullyVisibleRow(),
          });
        });
      });
    }
  };

  private generateWidget(
    className: string,
    annotation: string | JSX.Element,
    content: JSX.Element,
    addAsElementToScrollSync?: boolean
  ) {
    return (
      <div className={`av-widget ${className}`}>
        <div className="av-annotation">{annotation}</div>
        <div
          className="av-content"
          ref={(e) => {
            if (e && addAsElementToScrollSync) {
              //TODO: move into separate component .. Ref can be null here and
              //      also good to keep track of removal / addition for scroll sync
              //console.log("the ref is:", e);
              let scrollSync = ScrollSync.getInstance();
              scrollSync.registerScroller(e, "horizontal", true);
            }
          }}
        >
          {content}
        </div>
      </div>
    );
  }

  protected windowDimensionsUpdated = () => {
    this.setState({
      windowWidth: window.innerWidth,
      windowHeight: window.innerHeight,
    });
  };

  componentDidMount() {
    window.addEventListener("resize", this.windowDimensionsUpdated);
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.windowDimensionsUpdated);
  }

  render() {
    if (!this.props.alignment) {
      return null;
    }

    const annotationClass = this.props.showAnnotations
      ? ""
      : " annotation-closed";

    return (
      <div className={`alignment_viewer${annotationClass}`}>
        {this.renderMiniMap()}
        {/*<div id="column_mouseover"></div>*/}

        {this.generateWidget(
          "av-conservation-gaps",
          "Conservation / gaps:",
          this.renderConservationBox(),
          true
        )}

        {this.generateWidget(
          "av-sequence-logo",
          "Logo:",
          this.renderSequenceLogo(),
          true
        )}

        {this.generateWidget(
          "av-consensus-seq",
          "Consensus sequence:",
          this.renderConsensusQueryBox()
        )}

        {this.generateWidget(
          "av-target-seq",
          "Query sequence:",
          this.renderQuerySeqBox()
        )}

        {this.generateWidget(
          "av-position-indicator",
          "Position:",
          this.renderPositionBox()
        )}

        {this.generateWidget(
          "av-ace-msa",
          this.renderAlignmentAnnotationBox(),
          this.renderAlignmentBox()
        )}

        {/*
        //this is where we would add the webgl msa ..
        this.generateWidget(
          "av-webgl-msa",
          "",
          this.renderDatatableBox()
        )*/}
      </div>
    );
  }

  protected renderSequenceLogo = () => (
    <div
      className={
        `logo_box ${this.props.style.alignmentType.className} ` +
        `${this.props.style.colorScheme.className} ` +
        `${PositionsToStyle.ALL.className}`
      }
    >
      {
        <SequenceLogoComponent
          id="sequence_logo"
          alignment={this.props.alignment}
          glyphWidth={this.state.aceCharacterWidth}
          logoType={this.props.logoPlotStyle}
        />
      }
    </div>
  );

  protected renderConservationBox = () => (
    <SequenceConservationComponent
      id="sequence_conservation"
      alignment={this.props.alignment}
      characterWidth={this.state.aceCharacterWidth}
    />
  );

  protected renderConsensusQueryBox = () => (
    <div className="consensusseq_box">
      <AceConsensusSequenceComponent
        id="ace-consensusseq"
        alignment={this.props.alignment}
        fontSize={this.props.zoomLevel}
        sortBy={this.props.sortBy}
        classNames={[
          this.props.style.residueDetail.className,
          this.props.style.alignmentType.className,
          this.props.style.positionsToStyle.className,
          this.props.style.colorScheme.className,
        ].join(" ")}
        editorLoaded={(editor) => {
          this.aceEditorLoaded(
            "ace-consensusseq",
            editor,
            ScrollType.horizontal
          );
        }}
      ></AceConsensusSequenceComponent>
    </div>
  );

  protected renderQuerySeqBox = () => (
    <div className="queryseq_box">
      <AceTargetSequenceComponent
        id="ace-queryseq"
        alignment={this.props.alignment}
        fontSize={this.props.zoomLevel}
        sortBy={this.props.sortBy}
        classNames={[
          this.props.style.residueDetail.className,
          this.props.style.alignmentType.className,
          this.props.style.positionsToStyle.className,
          this.props.style.colorScheme.className,
        ].join(" ")}
        editorLoaded={(editor) => {
          this.aceEditorLoaded("ace-queryseq", editor, ScrollType.horizontal);
        }}
      ></AceTargetSequenceComponent>
    </div>
  );

  protected renderPositionBox = () => (
    <div className="position_box">
      {
        <AceTextualRulerComponent
          id="ace-positions"
          alignment={this.props.alignment}
          fontSize={this.props.zoomLevel}
          editorLoaded={(editor) => {
            this.aceEditorLoaded(
              "ace-positions",
              editor,
              ScrollType.horizontal
            );
          }}
          mouseEnter={() => {
            //console.log("mouse entered position box");
          }}
          mouseLeave={() => {
            //console.log("mouse left position box");
          }}
          mouseMove={(e) => {
            //console.log("mouse moved in position box", e);
          }}
          mouseDown={(e) => {
            //console.log("mouse down in position box", e);
          }}
        />
      }
    </div>
  );

  protected renderAlignmentBox = () => (
    <div className="alignment_box">
      <AceMultipleSequenceAlignmentComponent
        id="ace-alignment"
        alignment={this.props.alignment}
        fontSize={this.props.zoomLevel}
        sortBy={this.props.sortBy}
        classNames={[
          this.props.style.residueDetail.className,
          this.props.style.alignmentType.className,
          this.props.style.positionsToStyle.className,
          this.props.style.colorScheme.className,
        ].join(" ")}
        characterSizeChanged={this.handleCharacterSizeChanged}
        editorLoaded={(editor) => {
          this.aceEditorLoaded("ace-alignment", editor, ScrollType.both);
        }}
      ></AceMultipleSequenceAlignmentComponent>
    </div>
  );

  protected renderAlignmentAnnotationBox = () => (
    <div className="alignment_metadata_box">
      <AceEditorComponent
        id="ace-alignment-metadata"
        text={this.props.alignment
          .getSequences(this.props.sortBy)
          .map((x) => x.id)
          .join("\n")}
        fontSize={this.props.zoomLevel}
        editorLoaded={(editor) => {
          this.aceEditorLoaded(
            "ace-alignment-metadata",
            editor,
            ScrollType.vertical
          );
        }}
      ></AceEditorComponent>
    </div>
  );

  protected renderMiniMap() {
    const { alignment, sortBy, style, showMiniMap } = this.props;
    const { windowWidth, windowHeight } = this.state;

    const mmClassName = showMiniMap ? "mini-map" : "mini-map hidden";
    return (
      alignment &&
      style && (
        <div className={mmClassName}>
          <MiniMapComponent
            width={Math.min(400, alignment.getMaxSequenceLength())}
            height={windowHeight}
            alignHorizontal={"right"}
            alignment={alignment}
            style={style}
            sortBy={sortBy}
            highlightRows={
              this.state.alignmentEditorVisibleFirstRow !== undefined &&
              this.state.alignmentEditorVisibleLastRow !== undefined
                ? [
                    this.state.alignmentEditorVisibleFirstRow,
                    this.state.alignmentEditorVisibleLastRow,
                  ]
                : undefined
            }
            onClick={this.onMinimapClick}
          />
        </div>
      )
    );
  }

  protected onMinimapClick = (x: number, y: number) => {
    const {
      aceEditors,
      alignmentEditorVisibleFirstRow: alignmentEditorFirstRow,
      alignmentEditorVisibleLastRow: alignmentEditorLastRow,
    } = this.state;
    let rowCount = 40;
    if (
      alignmentEditorFirstRow !== undefined &&
      alignmentEditorLastRow !== undefined
    ) {
      rowCount = alignmentEditorLastRow - alignmentEditorFirstRow;
    }

    if (aceEditors.length >= 1) {
      aceEditors[0].scrollToRow(Math.floor(y - rowCount / 2));
    }
  };
}
