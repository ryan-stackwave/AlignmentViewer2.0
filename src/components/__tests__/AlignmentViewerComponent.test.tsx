import "jest-webgl-canvas-mock";
import * as React from "react";

import { shallow } from "enzyme";

import { Alignment, SequenceSortOptions } from "../../common/Alignment";
import { AlignmentViewer } from "../AlignmentViewerComponent";
import { AlignmentStyle, AlignmentTypes } from "../../common/MolecularStyles";
import { LOGO_TYPES } from "../SequenceLogoComponent";

// Due to the runtime necessities of using styles, we need to explicitly mock out some stub data.
// https://github.com/facebook/jest/issues/3094
jest.mock("../MolecularStyles.module.scss", () => {
  return {
    aaStyBGAlpha_Default: "",
    aaStyClass_Default: "",
    aaStyColorOrder_Default: "",
    aaStyColors_Default: "",
    aaStyDesc_Default: "",
    ntStyBGAlpha_Default: "",
    ntStyClass_Default: "",
    ntStyColorOrder_Default: "",
    ntStyColors_Default: "",
    ntStyDesc_Default: "",
  };
});

describe("AlignmentViewer", () => {
  it("Should render", () => {
    const wrapper = shallow(
      <AlignmentViewer
        alignment={new Alignment("", [])}
        style={AlignmentStyle.fromAlignmentType(AlignmentTypes.NUCLEOTIDE)}
        zoomLevel={0}
        sortBy={SequenceSortOptions.INPUT}
        logoPlotStyle={LOGO_TYPES.BARS}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
