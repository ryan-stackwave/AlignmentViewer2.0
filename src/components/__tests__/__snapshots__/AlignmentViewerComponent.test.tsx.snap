// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlignmentViewer Should render 1`] = `
<div
  className="alignment_viewer annotation-closed"
>
  <div
    className="mini-map hidden"
  >
    <MiniMapComponent
      alignHorizontal="right"
      alignment={
        Alignment {
          "aceEditorMode": undefined,
          "allAlphaLettersSorted": Array [],
          "consensus": Array [],
          "globalAlphaLetterCounts": Object {},
          "name": "",
          "positionalLetterCounts": Map {},
          "sequences": Map {
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            } => Array [],
          },
          "targetSequence": undefined,
          "uuid": "10000000-1000-4000-8000-100000000000",
        }
      }
      height={768}
      onClick={[Function]}
      sortBy={
        SequenceSortOptions {
          "description": "As input",
          "distanceFn": undefined,
          "isTargetSequenceDependent": false,
          "key": "as-input",
        }
      }
      style={
        NucleotideAlignmentStyle {
          "alignmentType": AlignmentTypes {
            "allColorSchemes": Array [
              Object {
                "backgroundAlpha": NaN,
                "className": "",
                "colors": Object {
                  "": "",
                },
                "commonName": "Default",
                "description": "",
              },
            ],
            "className": undefined,
            "description": "Nucleotide Sequences",
            "key": "nucleotide-alignment",
          },
          "allColorSchemes": Array [
            Object {
              "backgroundAlpha": NaN,
              "className": "",
              "colors": Object {
                "": "",
              },
              "commonName": "Default",
              "description": "",
            },
          ],
          "colorScheme": Object {
            "backgroundAlpha": NaN,
            "className": "",
            "colors": Object {
              "": "",
            },
            "commonName": "Default",
            "description": "",
          },
          "positionsToStyle": PositionsToStyle {
            "className": undefined,
            "description": "All",
            "key": "all",
          },
          "residueDetail": ResidueDetailTypes {
            "className": "",
            "description": "Style Background and Letters",
            "key": "both",
          },
        }
      }
      width={-1}
    />
  </div>
  <div
    className="av-widget av-conservation-gaps"
  >
    <div
      className="av-annotation"
    >
      Conservation / gaps:
    </div>
    <div
      className="av-content"
    >
      <SequenceConservationComponent
        alignment={
          Alignment {
            "aceEditorMode": undefined,
            "allAlphaLettersSorted": Array [],
            "consensus": Array [],
            "globalAlphaLetterCounts": Object {},
            "name": "",
            "positionalLetterCounts": Map {},
            "sequences": Map {
              SequenceSortOptions {
                "description": "As input",
                "distanceFn": undefined,
                "isTargetSequenceDependent": false,
                "key": "as-input",
              } => Array [],
            },
            "targetSequence": undefined,
            "uuid": "10000000-1000-4000-8000-100000000000",
          }
        }
        characterWidth={0}
        id="sequence_conservation"
      />
    </div>
  </div>
  <div
    className="av-widget av-sequence-logo"
  >
    <div
      className="av-annotation"
    >
      Logo:
    </div>
    <div
      className="av-content"
    >
      <div
        className="logo_box undefined  undefined"
      >
        <SequenceLogoComponent
          alignment={
            Alignment {
              "aceEditorMode": undefined,
              "allAlphaLettersSorted": Array [],
              "consensus": Array [],
              "globalAlphaLetterCounts": Object {},
              "name": "",
              "positionalLetterCounts": Map {},
              "sequences": Map {
                SequenceSortOptions {
                  "description": "As input",
                  "distanceFn": undefined,
                  "isTargetSequenceDependent": false,
                  "key": "as-input",
                } => Array [],
              },
              "targetSequence": undefined,
              "uuid": "10000000-1000-4000-8000-100000000000",
            }
          }
          glyphWidth={0}
          id="sequence_logo"
          logoType="bar plot"
        />
      </div>
    </div>
  </div>
  <div
    className="av-widget av-consensus-seq"
  >
    <div
      className="av-annotation"
    >
      Consensus sequence:
    </div>
    <div
      className="av-content"
    >
      <div
        className="consensusseq_box"
      >
        <AceConsensusSequenceComponent
          alignment={
            Alignment {
              "aceEditorMode": undefined,
              "allAlphaLettersSorted": Array [],
              "consensus": Array [],
              "globalAlphaLetterCounts": Object {},
              "name": "",
              "positionalLetterCounts": Map {},
              "sequences": Map {
                SequenceSortOptions {
                  "description": "As input",
                  "distanceFn": undefined,
                  "isTargetSequenceDependent": false,
                  "key": "as-input",
                } => Array [],
              },
              "targetSequence": undefined,
              "uuid": "10000000-1000-4000-8000-100000000000",
            }
          }
          classNames="   "
          editorLoaded={[Function]}
          fontSize={0}
          id="ace-consensusseq"
          sortBy={
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="av-widget av-target-seq"
  >
    <div
      className="av-annotation"
    >
      Query sequence:
    </div>
    <div
      className="av-content"
    >
      <div
        className="queryseq_box"
      >
        <AceTargetSequenceComponent
          alignment={
            Alignment {
              "aceEditorMode": undefined,
              "allAlphaLettersSorted": Array [],
              "consensus": Array [],
              "globalAlphaLetterCounts": Object {},
              "name": "",
              "positionalLetterCounts": Map {},
              "sequences": Map {
                SequenceSortOptions {
                  "description": "As input",
                  "distanceFn": undefined,
                  "isTargetSequenceDependent": false,
                  "key": "as-input",
                } => Array [],
              },
              "targetSequence": undefined,
              "uuid": "10000000-1000-4000-8000-100000000000",
            }
          }
          classNames="   "
          editorLoaded={[Function]}
          fontSize={0}
          id="ace-queryseq"
          sortBy={
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="av-widget av-position-indicator"
  >
    <div
      className="av-annotation"
    >
      Position:
    </div>
    <div
      className="av-content"
    >
      <div
        className="position_box"
      >
        <AceTextualRulerComponent
          alignment={
            Alignment {
              "aceEditorMode": undefined,
              "allAlphaLettersSorted": Array [],
              "consensus": Array [],
              "globalAlphaLetterCounts": Object {},
              "name": "",
              "positionalLetterCounts": Map {},
              "sequences": Map {
                SequenceSortOptions {
                  "description": "As input",
                  "distanceFn": undefined,
                  "isTargetSequenceDependent": false,
                  "key": "as-input",
                } => Array [],
              },
              "targetSequence": undefined,
              "uuid": "10000000-1000-4000-8000-100000000000",
            }
          }
          editorLoaded={[Function]}
          fontSize={0}
          id="ace-positions"
          mouseDown={[Function]}
          mouseEnter={[Function]}
          mouseLeave={[Function]}
          mouseMove={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="av-widget av-ace-msa"
  >
    <div
      className="av-annotation"
    >
      <div
        className="alignment_metadata_box"
      >
        <AceEditorComponent
          editorLoaded={[Function]}
          fontSize={0}
          id="ace-alignment-metadata"
          text=""
        />
      </div>
    </div>
    <div
      className="av-content"
    >
      <div
        className="alignment_box"
      >
        <AceMultipleSequenceAlignmentComponent
          alignment={
            Alignment {
              "aceEditorMode": undefined,
              "allAlphaLettersSorted": Array [],
              "consensus": Array [],
              "globalAlphaLetterCounts": Object {},
              "name": "",
              "positionalLetterCounts": Map {},
              "sequences": Map {
                SequenceSortOptions {
                  "description": "As input",
                  "distanceFn": undefined,
                  "isTargetSequenceDependent": false,
                  "key": "as-input",
                } => Array [],
              },
              "targetSequence": undefined,
              "uuid": "10000000-1000-4000-8000-100000000000",
            }
          }
          characterSizeChanged={[Function]}
          classNames="   "
          editorLoaded={[Function]}
          fontSize={0}
          id="ace-alignment"
          sortBy={
            SequenceSortOptions {
              "description": "As input",
              "distanceFn": undefined,
              "isTargetSequenceDependent": false,
              "key": "as-input",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
